### ISSUE TITLE
'unable to fetch client for account ID' error when creating role under AWS mount with IAM role from another account

### ISSUE THREAD

**Issue by EXPEddrewery** [hashicorp/vault#3681]
When using Vault 0.8.2, we are finding that the retrieval of the stsRole using the stsRoleForAccount function in client.go (below) is returning "" when the value is present.

https://github.com/hashicorp/vault/blob/c641938cefd374811bc1afdbb1241e74947ad431/builtin/credential/aws/client.go#L227

I'm 99% sure this is where the issue is occurring as I can't see any other code paths that would lead to the error other than that returning "". Other paths would result in different errors.

We noticed this when we were testing some cross account functionality where we are creating a new role using the `aws/role/{role_name}` endpoint and the IAM role is in a different account to Vault.

The STS role had been correctly configured using the "aws/config/sts" endpoint as shown by the below output (note I have removed actual account ids and roles):

```
vault read auth/aws/config/sts/{account_id}
Key     	Value
---     	-----
sts_role	arn:aws:iam::{account_id}:role/{cross_auth_role}
```

When we attempt to create the role, we receive the following :

```
vault write auth/aws/role/mytestrole1 bound_iam_principal_arn="arn:aws:iam::{account_id}:role/{executing_iam_role_name}" policies="{mypolicies}"
Error writing data to auth/aws/role/mytestrole1: Error making API request.

URL: PUT https://{vault_server}/v1/auth/aws/role/mytestrole1
Code: 400. Errors:

* failed updating the unique ID of ARN "arn:aws:iam::{account_id}:role/{executing_iam_role_name}": &errors.errorString{s:"unable to fetch client for account ID {account_id} -- default client is for account {default_account_id}"}
```

After a few attempts over the next few seconds, it then succeeds :

```
vault write auth/aws/role/mytestrole1 bound_iam_principal_arn="arn:aws:iam::{account_id}:role/{executing_iam_role_name}" policies="{mypolicies}"
Success! Data written to: auth/aws/role/mytestrole1
```

It may be unrelated but some of our roles are dot separated (eg: "my.role-name"). I couldn't replicate it today while trying to get the information for this issue together, but after changing the role name from "mytestrole1" to "my.testrole1" it happened every second time, even after switching back to "mytestrole1".

I had a look through the code and couldn't see anything obvious so thought I would raise this issue for the experts to dig into it.

**Comment by joelthompson** [comment]
Hey @EXPEddrewery,

Agree, this sounds very strange, and I'm not seeing what could cause it. Can you describe your Vault setup in a bit more detail? E.g.:

* Which edition you're using, e.g., open source or a paid version, and if a paid version, are you using any replication?
* What storage backend you're using
* How many nodes in your Vault cluster
* Anything else non-standard you might be doing

You also said it happened every second time, but did it eventually "settle down" and start working consistently? And what is your pattern of writes to `auth/aws/config/sts/{account_id}`? Are there possibly some jobs that are fighting over it (e.g., one which deletes it and one which recreates it)?

It basically sounds like https://github.com/hashicorp/vault/blob/9afe7330e06e486ee326621624f2077d88bc9511/builtin/credential/aws/path_config_sts.go#L148-L155 is returning inconsistent values on each call. A few thoughts on what could cause it (and I'm mostly thinking out loud here; the storage layer isn't something I'm really familiar with):

* A bug in the Vault storage layer (but I'd expect more widespread bug reports)
* Your storage backend not supporting read-after-write consistency causing it to return inconsistent results (but I'd expect Vault's caching layer to take care of this as well as more widespread issues from your deployment)
* Some sort of replication lag if you're load balancing requests across performance-replicated Vault clusters in Vault Enterprise
* A non-supported custom storage replication with requests load balanced across Vault clusters pointed to the different replicated storage backends (and Vault's caching would explain this)
* Different jobs that are racing against each other trying to write to `auth/aws/config/sts/{account_id}`

The role name having a period should not matter at all, and every write to `auth/aws/role/<role_name>` specifying a `bound_iam_principal_arn` (that isn't suffixed with a `*`) should trigger the exact same code path in question here (intentionally, so as to allow refreshing the internally stored AWS unique ID).

**Comment by EXPEddrewery** [comment]
Amazing timing @joelthompson! 

We just figured out (literally 5 minutes ago) that it is due to us having two AWS auth backends (one we use as a fallback mechanism for our deployments).

We are configuring the primary backend with the STS configuration for each of the accounts, but not the fallback.

I've just added the configuration to the fallback and it is working as expected.

My sincere apologies! It certainly explains why the code looked fine... because it is!

Let's chalk this one up to PEBKAC.
